_posts/2020-03-06-ok-mauro-but-i-want-to-do-paging-and-sorting.md

```diff
@@ -0,0 +1,42 @@
+---
+layout: post
+header_image: /img/posts/ok-mauro-but-i-want-to-do-paging-and-sorting/header.jpg
+title: "OK Mauro, but I want to do paging AND sorting"
+author: Mauro Servienti
+synopsis: "Paging and sorting can be easily applied in isolation, even in distributed systems. When in need of paging AND sorting, the level of coordination required grows exponentially. It turns out that the problem is not that different from searching in distributed systems, and similar techniques can be used to address the paging and sorting dilemma."
+tags:
+- SOA
+- Services ViewModel Composition
+category: view-model-composition
+---
+
+It's a kind of [tornado of souls](https://www.youtube.com/watch?v=K-HzFACAedk) here. Coronavirus is all around where I live, and smart working, as Italians call remote working, is all the rage. Articles, posts, and twitter threads are all about how to be great at remote working.
+
+One way or the other, I've been working remotely for the last ten years or so, and in the last six years, I've been 100% remote working.
+Don't worry. I don't want to talk "remote working," at least not today.
+
+This post assumes that the ["Paging and sorting in distributed systems, oh my!"](https://milestone.topics.it/view-model-composition/2020/01/27/paging-and-sorting-in-distributed-systems-oh-my.html) article, has been read. If this is not the case, go and read it. I'll wait here.
+
+As you probably noticed, the approaches described to page data and sort data work in isolation, but immediately fall apart if we need to use them together. If the need is to apply "paging and sorting" in the same request, then the proposed solutions cannot be applied. Or, if implemented, can only result in a very chatty and performance-avert design.
+
+## Paging and sorting is misleading
+
+Data sorting needs to happen before paging can be applied. If we page and then sort, we're going to sort only the current page, which means that if the user transitions to another page, the sort doesn't make sense anymore. It should be "sorting and paging," but that sounds bad.
+
+The problem is that if we have to sort first and the request coming in is like `/products?orderBy=price,description` then, as we already said, sorting is not so easy. The request reveals a relationship between the service owning prices and the service owning descriptions. If data are not in the same storage, the only option if first to load all the prices and sort them, then handover the result set to the service that can load all the descriptions and sort the result set based on price and description. Only at this point paging can be applied.
+
+## That's a join, Mauro
+
+If we were to apply the outlined logic manually, it's like we're reinventing the wheel. And we've already discussed something very similar, lately.
```

Comment:
```suggestion
If we were to apply the outlined logic manually, it's like we're reinventing the wheel. And we've already discussed something very similar recently.
```

Maybe a link to the post where it's discussed. Better yet, a link to the heading in that post.